### PR TITLE
Ping pong with RTT

### PR DESCRIPTION
**Pull request by mirpo**

### CODE REVIEW

File: src/connection/WsEndpoint.js [hunk 1]
```diff
@@ -135,22 +135,41 @@ class WsEndpoint extends EventEmitter {
...
-                ws.ping('ping')
+                // didn't get "pong" in pingInterval
+                if (ws.respondedPong !== undefined && !ws.respondedPong) {
+                    throw Error('ws is not active')
+                }
+
+                // eslint-disable-next-line no-param-reassign
+                ws.respondedPong = false
+                ws.rttStart = Date.now()
+                ws.rtt = ws.rttStart - ws.rttEnd
```
Comment (harbu) [inline src/connection/WsEndpoint.js]: Shouldn't this be moved to `pong`?

```js
pong: (ws) => {
                const connection = this.connections.get(ws.address)

                if (connection) {
                    this.debug(`received from ${ws.address} "pong" frame`)
                    connection.respondedPong = true
                    connection.rtt = Date.now() - connection.rttStart
                }
```

File: src/connection/WsEndpoint.js [hunk 2]
```diff
@@ -477,6 +497,14 @@ class WsEndpoint extends EventEmitter {
             setImmediate(() => this.onReceive(peerInfo, address, message))
         })
 
+        ws.on('ping', () => {
```
Comment (harbu) [inline src/connection/WsEndpoint.js]: Don't think this is needed though? Why should a received ping affect this node's RTT calculation? I think the RTT calculation from node A to node B should be independent from calculation from node B to node A.

File: src/logic/Node.js [hunk 3]
```diff
@@ -342,13 +342,16 @@ class Node extends EventEmitter {
     _sendStreamStatus(streamId) {
         const streamKey = streamId.key()
         const trackerId = this.trackersRing.get(streamKey)
-        clearTimeout(this.sendStatusTimeout.get(trackerId))
 
-        const timeout = setTimeout(() => {
-            this._sendStatus(trackerId)
-        }, this.opts.sendStatusToAllTrackersInterval)
+        if (trackerId) {
```
Comment (harbu) [inline src/logic/Node.js:346]: Don't understand? When would a trackerId not be returned for a given streamKey? It always has to be assigned to some tracker?

File: src/connection/WsEndpoint.js (hunk 1 above)
Comment (harbu) [reply src/connection/WsEndpoint.js]: Otherwise we only call setInterval() every `pingInterval` so this will lead to error in measurement?

**Review by harbu — CHANGES_REQUESTED** [review]

File: src/logic/Node.js (hunk 3 above)
Comment (mirpo) [reply src/logic/Node.js:346]: for example:
```js
startNetworkNode(program.ip, program.port, id).then((subscriber) => {
    console.log('started subscriber id: %s, port: %d, ip: %s, trackers: %s, streamId: %s, metrics: %s',
        id, program.port, program.ip, program.trackers.join(', '), program.streamId, program.metrics)

    subscriber.subscribe(program.streamId, 0)
    program.trackers.map((trackerAddress) => subscriber.addBootstrapTracker(trackerAddress))
```

we subscribe, but not connected yet

File: src/connection/WsEndpoint.js (hunk 1 above)
Comment (mirpo) [reply src/connection/WsEndpoint.js]: Good point

File: src/connection/WsEndpoint.js (hunk 2 above)
Comment (mirpo) [reply src/connection/WsEndpoint.js]: We have two kinds of connections -> inbound and outbound, that's we sometimes we ping `uws` socket, sometimes `ws`. So uws `pong` will be arrive to `ws server`, `ws` will be triggered in this event.

File: src/logic/Node.js (hunk 3 above)
Comment (harbu) [reply src/logic/Node.js:346]: Alright so basically when there are no trackers added.

File: src/connection/WsEndpoint.js (hunk 2 above)
Comment (harbu) [reply src/connection/WsEndpoint.js]: But is this event triggered when we _receive_ a ping? The only thing I think that should happen when receiving ping is to respond pong.

1. Node A: record `startTime`, send `ping` to Node B
2. Node B: receive `ping` from Node A, send `pong` to Node A
3. Node A: receive `pong` from Node B, calculate `rtt = startTime - now()`

**Review by juslesan — APPROVED** [review]

**Review by harbu — APPROVED** [review]
LGTM!

**Pull request merged**